src/writers.rs

```diff
@@ -0,0 +1,381 @@
+use crate::column_type::*;
+use crate::statements::PyPreparedStatement;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::{Bound, IntoPyObjectExt, PyResult, Python, pyclass, pymethods, pymodule};
+use scylla::_macro_internal::{CellWriter, ColumnType, RowSerializationContext};
+use scylla::cluster::metadata::{CollectionType, NativeType};
+use scylla::serialize::SerializationError;
+use scylla::serialize::writers::CellOverflowError;
+use scylla::statement::prepared::PreparedStatement;
+use scylla_cql::serialize::row::SerializedValues;
+use std::sync::{Arc, Mutex};
+
+pub enum PySerializationError {
+    Serialization(SerializationError),
+    Overflow(CellOverflowError),
+}
+
+impl From<PySerializationError> for PyErr {
+    fn from(err: PySerializationError) -> Self {
+        PyValueError::new_err(match err {
+            PySerializationError::Serialization(e) => e.to_string(),
+            PySerializationError::Overflow(e) => e.to_string(),
+        })
+    }
+}
+
+impl From<SerializationError> for PySerializationError {
+    fn from(err: SerializationError) -> Self {
+        PySerializationError::Serialization(err)
+    }
+}
+
+impl From<CellOverflowError> for PySerializationError {
+    fn from(err: CellOverflowError) -> Self {
+        PySerializationError::Overflow(err)
+    }
+}
+
+#[pyclass]
+pub struct PyRowSerializationContext {
+    prepared_statement: Arc<PreparedStatement>,
+}
+
+#[pymethods]
+impl PyRowSerializationContext {
+    #[staticmethod]
+    pub fn from_prepared(prepared_statement: PyRef<'_, PyPreparedStatement>) -> Self {
+        Self {
+            prepared_statement: Arc::clone(&prepared_statement._inner),
+        }
+    }
+
+    pub fn column_count(&self) -> usize {
+        RowSerializationContext::from_specs(
+            self.prepared_statement.get_variable_col_specs().as_slice(),
+        )
+        .columns()
+        .len()
+    }
+
+    pub fn get_columns(&self) -> PyResult<Vec<Py<PyAny>>> {
+        let context = RowSerializationContext::from_specs(
+            self.prepared_statement.get_variable_col_specs().as_slice(),
+        );
+        let mut columns = Vec::new();
+
+        Python::with_gil(|py| {
+            for col_spec in context.columns() {
+                let typ = extract_type(py, col_spec.typ())?;
+
+                columns.push(typ);
+            }
+
+            Ok(columns)
+        })
+    }
+}
+
```

Comment:
This all should be precomputed on construction, not computed on each call to `get_columns`.